src/connection.rs

```diff
@@ -92,438 +92,322 @@ impl<T: AsyncRead + AsyncWrite> Connection<T> {
-    pub fn set_max_buffer_size(&mut self, max: usize) -> &mut Self {
-        self.max_buffer_size = max;
+    /// Message fragments will be buffered and concatenated up to this value.
+    pub fn set_max_message_size(&mut self, max: usize) -> &mut Self {
+        self.max_message_size = max;
         self
     }
 
-    fn set_mask(&self, frame: &mut Frame) {
-        if self.mode.is_client() {
-            frame.header_mut().set_masked(true);
-            frame.header_mut().set_mask(rand::thread_rng().next_u32());
-        }
+    /// Toggle UTF-8 check for incoming text messages.
+    pub fn validate_utf8(&mut self, value: bool) -> &mut Self {
+        self.validate_utf8 = value;
+        self
     }
-}
 
-impl<T: AsyncRead + AsyncWrite> Connection<T> {
-    fn answer_ping(&mut self, frame: Frame, buf: Option<base::Data>) -> Poll<(), Error> {
-        trace!("answering ping: {:?}", frame.header());
-        if let AsyncSink::NotReady(frame) = self.framed.start_send(frame)? {
-            self.state = Some(State::AnswerPing(frame, buf));
-            return Ok(Async::NotReady)
-        }
-        self.flush(buf)
+    /// Send some binary data over this connection.
+    pub async fn send_binary(&mut self, data: &mut BytesMut) -> Result<(), Error> {
+        let mut header = Header::new(OpCode::Binary);
+        self.send(&mut header, data).await?;
+        Ok(())
     }
 
-    fn answer_close(&mut self, frame: Frame) -> Poll<(), Error> {
-        trace!("answering close: {:?}", frame.header());
-        if let AsyncSink::NotReady(frame) = self.framed.start_send(frame)? {
-            self.state = Some(State::AnswerClose(frame));
-            return Ok(Async::NotReady)
-        }
-        self.closing()
+    /// Send some text data over this connection.
+    pub async fn send_text(&mut self, data: &mut BytesMut) -> Result<(), Error> {
+        debug_assert!(std::str::from_utf8(&data).is_ok());
+        let mut header = Header::new(OpCode::Text);
+        self.send(&mut header, data).await?;
+        Ok(())
     }
 
-    fn send_close(&mut self, frame: Frame) -> Poll<(), Error> {
-        trace!("sending close: {:?}", frame.header());
-        if let AsyncSink::NotReady(frame) = self.framed.start_send(frame)? {
-            self.state = Some(State::SendClose(frame));
-            return Ok(Async::NotReady)
+    /// Flush the socket buffer.
+    pub async fn flush(&mut self) -> Result<(), Error> {
+        trace!("flushing connection");
+        if self.is_closed {
+            return Ok(())
         }
-        self.flush_close()
-    }
-
-    fn flush_close(&mut self) -> Poll<(), Error> {
-        trace!("flushing close");
-        if self.framed.poll_complete()?.is_not_ready() {
-            self.state = Some(State::FlushClose);
-            return Ok(Async::NotReady)
+        if !self.wbuffer.is_empty() {
+            self.socket.write_all(&self.wbuffer).await?;
+            trace!("flushed {} bytes", self.wbuffer.len());
```

Comment:
```suggestion
            trace!("wrote {} bytes", self.wbuffer.len());
```
A semantic nitpick, but I think the data may not actually have been flushed yet at this point, only after `self.socket.flush()`.

Reply:
The "flushed" is referring to the method where the `trace!` is located in and the wording is a deliberate attempt to distinguish it from the "wrote" that is used in [`Connection::write`](https://github.com/paritytech/soketto/blob/e25b0a18837c676a2e97f8eff3238061dc18b3a7/src/connection.rs#L190).